build.assets/changelog.sh

```diff
@@ -1,15 +1,204 @@
+		return 1
+	}
+	e_time=$(git log -n 1 --date=format:'%Y-%m-%dT%H:%M:%S%z' --format=%cd e) || {
+		echo 'Cant get last modified time of "e" ref' >&2
+		return 1
+	}
+
+	list_prs "${branch}" "${since}"
+	printf '\nEnterprise:\n'
+	(cd e && list_prs "${branch}" "${since_e}" "${e_time}")
+
+	return 0
+}
+
+check_prereq() {
+	if command -v jq >/dev/null 2>&1; then
+		JQ=jq
+		return 0
+	fi
+	if command -v gojq >/dev/null 2>&1; then
+		JQ=gojq
+		return 0
+	fi
+
+	echo 'jq or gojq not installed. Install gojq easily with:' >&2
+	echo 'go install github.com/itchyny/gojq/cmd/gojq@latest' >&2
+	return 1
+}
+
+get_branch() {
+	# If BASE_BRANCH is set, just use that. Otherwise try to figure it out.
+	if [[ -n "${BASE_BRANCH-}" ]]; then
+		echo "${BASE_BRANCH}"
+		return 0
+	fi
+
+	local ref branch
+	ref=$(git symbolic-ref HEAD 2>/dev/null) || {
+		echo 'Not on a branch' >&2
+		return 1
+	}
+	branch="${ref#refs/heads/}"
+	if [[ "${ref}" == "${branch}" ]]; then
+		echo "Not on a branch: ${ref}" >&2
+		return 1
+	fi
+
+	if [[ "${branch}" != branch/v* ]]; then
+		# If we're already on the branch cut for the release, try to
+		# determine the root branch name
+		local fbranch
+		fbranch=$(
+			git branch \
+				--list 'branch/v*' \
+				--contains "$(git merge-base --fork-point HEAD)" \
+				--format '%(refname:short)'
+		)
+		branch="${fbranch:-${branch}}" # Don't overwrite $branch with empty
+	fi
+	if ! [[ "${branch}" == branch/v* ]]; then
+		echo "Not on a release branch: ${branch}" >&2
+		return 1
+	fi
+
+	echo "${branch}"
+	return 0
+}
+
+get_last_version() {
+	# If BASE_TAG is set, just use that, otherwise figure out the last version
+	if [[ -n "${BASE_TAG-}" ]]; then
+		echo "${BASE_TAG}"
+		return 0
+	fi
+
+	cd "$(git rev-parse --show-toplevel 2>/dev/null)" || {
+		echo 'ugh. cant cd to repo root' >&2
+		return 1
+	}
+	local last_version
+	last_version=$(make -s print-version) || {
+		echo 'Cant get last released version' >&2
+		return 1
+	}
+
+	echo "v${last_version}"
+	return 0
+}
+
+list_prs() {
+	local branch="$1" from="$2" to="${3-}"
+	local merged_query
+	if [[ -z "${to}" ]]; then
+		merged_query="merged:>${from}"
+	else
+		merged_query="merged:${from}..${to}"
+	fi
+
+	local gh_query="base:${branch} ${merged_query} -label:no-changelog"
 
-COMMIT=$(git rev-list -n 1 v$BASE_TAG)
+	# shellcheck disable=SC2016 # We're not trying to expand in single quotes
+	jq_expr='
+	  def extract_cl: gsub("\r"; "") | split("\n") | [.[] | scan("^[Cc]hangelog: +(.*)$") | .[]];
+	  def promote_title: {number, url, changelog: (if .changelog == [] then ["NOCL: " + .title] else .changelog end)};
+	  def flatten_changes: . as $pr | .changelog[] | $pr + {changelog: .};
+	  def clean: sub("\\s*$"; "") | if . | endswith(".") then . else . + "." end;
+	  def as_entry: "* \(.changelog | clean) [#\(.number)](\(.url))";
+	  map({number, url, title, changelog: .body | extract_cl}) |
+	  map(promote_title) |
+	  map(flatten_changes) |
+	  map(as_entry) |
+	  join("\n")
```

Comment:
Wow, we are balancing on a thin line between sanity and using esoteric programming languages here. It's still probably better than using Perl, though. ;)

I guess it's inevitable, I myself don't see a better way to do it, but just wanted to let you know that now I understand why you wanted to rewrite this entire thing in Go. :)

Reply:
This was my personal changelog script that I was requested to add to the repo, so you all get it in its obtuse glory.

I want to make this do more - such as traversing backport PRs back to the original to get the changelog entry from there, and at that point I'm certainly not doing it in shell/jq. The current version at least has the simplicity of an immutable single pipeline - a github query gives a stream of PR details as json which is processed one at a time to transform into a set of changelog entries. Once we start querying other PRs in the middle of this, it's time for some proper code.